query/src/sql/executor/physical_plan_builder.rs

```diff
@@ -45,23 +47,26 @@ use crate::sql::plans::RelOperator;
 use crate::sql::plans::Scalar;
 use crate::sql::MetadataRef;
 use crate::sql::ScalarExpr;
+use crate::storages::ToReadDataSourcePlan;
 
 pub struct PhysicalPlanBuilder {
     metadata: MetadataRef,
+    ctx: Arc<QueryContext>,
 }
 
 impl PhysicalPlanBuilder {
-    pub fn new(metadata: MetadataRef) -> Self {
-        Self { metadata }
+    pub fn new(metadata: MetadataRef, ctx: Arc<QueryContext>) -> Self {
+        Self { metadata, ctx }
     }
 
-    pub fn build(&self, s_expr: &SExpr) -> Result<PhysicalPlan> {
+    #[async_recursion::async_recursion]
+    pub async fn build(&self, s_expr: &SExpr) -> Result<PhysicalPlan> {
```

Comment:
I'm not sure if it's a good idea to make `PhysicalPlanBuilder` async.

I stored `ReadDataSourcePlan` inside `Metadata` on purpose to prevent async function invocation in `PhysicalPlanBuilder`.

Reply:
The previous `ReadDataSourcePlan ` inside `Metadata` is an insufficient object.

It lakes `projection`, `predicate`, and `limit` pushdown that could result to the whole parts.

Reply:
@sundy-li Ok, let's accept it.